### DISCUSSION TITLE
How to get the encoding of an input.http stream?

### DISCUSSION THREAD

**Discussion by maardsma** [savonet/liquidsoap#1882] [Q&A] [answered]
I am downloading two mp3 streams via http. I want to determine their samplerate and/or bitrate so I can stream that back out at the same encoding. (No reason to transcode up.) Does the Liquidsoap API allow me to obtain this info from the input streams?

**Comment by toots** [comment]
Hi!

If your intent is to restream without transcoding, the way to do it is via the raw ffmpeg format. Something like this:
```ruby
s = input.http("http://...")

# You need to know at least the format:
encoder = %ffmpeg(format="mp3", %audio.raw)

# Restream without re-encoding:
output.icecast(
  mount="...",
  ...,
  fallible=true,
  encoder,
  s
)
```

Here's a full example with a switch between two sources:
```ruby
s1 = input.http("https://supersoul.live:8020/SRRS320kbps")

s2 = input.http("https://supersoul.live:8020/SRRS128kbps")

b = blank()

b = ffmpeg.encode.audio(
  %ffmpeg(%audio(codec="libmp3lame", b="128k")),
  b
)

s = switch(id="switch", track_sensitive=false, [
  ({file.exists("/tmp/s1")},s1),
  ({file.exists("/tmp/s2")},s2),
  ({true},b)
])

e = %ffmpeg(format="mp3", %audio.copy)

output.harbor(id="harbor", mount="test", e, s)
```
You can switch between the two sources by creating/removing files at `/tmp/s1` and `/tmp/s2`.

Hope this helps!

**Comment by maardsma** [comment]
Cool thank you!

I'm actually combining two sources via pan and add, and streaming them out, so I do need to re-encode. I was hoping to query the samplerate of the input sources to determine output samplerate. Is that possible?

**Comment by toots** [comment, marked as answer]
Ha. Well, in this case, the data is converted to the internal format's samplerate, set via:
```ruby'
audio.samplerate.set(48000)
```
You will have to have all your incoming streams with the same samplerate if you want to avoid resampling.

As a rule of thumb, I would consider the trouble/benefit about tweaking this pretty minor in terms of CPU load and/or audio quality. However, if you really want to try to be conservative, you can set the internal samplerate to a the higest common value, for instance `48000` and trust conversion to preserve most information.

There are also specific settings that you can use with the `libsamplerate` based converter, some of which will give you higher quality resampling. However, they are pretty CPU-intensive.

**Reply by maardsma** [reply]
Ah ok this is a key concept, that there's one sample rate used during internal processing. I will set this to my planned outbound sample rate to save a bit of internal processing power.

I was able to view the samplerate of those streams on a one-time basis using QuickTime player. I don't need this behavior to be dynamic so that works.

Mark

**Reply by maardsma** [reply]
Is this setting done by set("frame.audio.samplerate", int) in version 1.4.4?

**Comment by toots** [comment]
Yes!
